File: frontend/lib/calculations/poisson.ts
import { PoissonPrediction } from '@/types/betting';

function factorial(n: number): number {
  if (n <= 1) return 1;
  let result = 1;
  for (let i = 2; i <= n; i++) {
    result *= i;
  }
  return result;
}

function poissonProbability(lambda: number, k: number): number {
  return (Math.pow(lambda, k) * Math.exp(-lambda)) / factorial(k);
}

export function calculatePoissonPrediction(
  homeGoalsAvg: number,
  homeConcededAvg: number,
  awayGoalsAvg: number,
  awayConcededAvg: number,
  leagueAvgGoals: number = 2.75
): PoissonPrediction {
  const homeAttackStrength = homeGoalsAvg / (leagueAvgGoals / 2);
  const homeDefenseStrength = homeConcededAvg / (leagueAvgGoals / 2);
  const awayAttackStrength = awayGoalsAvg / (leagueAvgGoals / 2);
  const awayDefenseStrength = awayConcededAvg / (leagueAvgGoals / 2);

  const expectedHomeGoals = homeAttackStrength * awayDefenseStrength * (leagueAvgGoals / 2);
  const expectedAwayGoals = awayAttackStrength * homeDefenseStrength * (leagueAvgGoals / 2);

  const maxGoals = 10;
  const scoreMatrix: number[][] = [];
  
  for (let i = 0; i <= maxGoals; i++) {
    scoreMatrix[i] = [];
    for (let j = 0; j <= maxGoals; j++) {
      scoreMatrix[i][j] = poissonProbability(expectedHomeGoals, i) * poissonProbability(expectedAwayGoals, j);
    }
  }

  let homeWinProb = 0;
  let drawProb = 0;
  let awayWinProb = 0;
  let over25Prob = 0;
  let bttsProb = 0;

  for (let i = 0; i <= maxGoals; i++) {
    for (let j = 0; j <= maxGoals; j++) {
      const prob = scoreMatrix[i][j];
      if (i > j) homeWinProb += prob;
      else if (i === j) drawProb += prob;
      else awayWinProb += prob;
      
      if (i + j > 2) over25Prob += prob;
      if (i > 0 && j > 0) bttsProb += prob;
    }
  }

  const mostLikelyScores: { score: string; probability: number }[] = [];
  for (let i = 0; i <= 5; i++) {
    for (let j = 0; j <= 5; j++) {
      mostLikelyScores.push({
        score: `${i}-${j}`,
        probability: scoreMatrix[i][j] * 100,
      });
    }
  }
  mostLikelyScores.sort((a, b) => b.probability - a.probability);

  return {
    homeGoals: expectedHomeGoals,
    awayGoals: expectedAwayGoals,
    homeWinProb: homeWinProb * 100,
    drawProb: drawProb * 100,
    awayWinProb: awayWinProb * 100,
    over25Prob: over25Prob * 100,
    under25Prob: (1 - over25Prob) * 100,
    bttsProb: bttsProb * 100,
    scoreMatrix,
    mostLikelyScores: mostLikelyScores.slice(0, 10),
  };
}

export function calculateMatchProbabilities(
  homeGoalsAvg: number,
  awayGoalsAvg: number
): { homeWin: number; draw: number; awayWin: number } {
  const prediction = calculatePoissonPrediction(
    homeGoalsAvg,
    2.5,
    awayGoalsAvg,
    2.5
  );
  
  return {
    homeWin: prediction.homeWinProb,
    draw: prediction.drawProb,
    awayWin: prediction.awayWinProb,
  };
}

File: backend/lib/calculations/poisson.go
// Package calculations provides mathematical models for betting analytics and stock valuation.
// These functions mirror the frontend TypeScript implementations in frontend/lib/calculations.
package calculations

import "math"

// Factorial calculates n! for Poisson calculations.
func Factorial(n int) float64 {
	if n <= 1 {
		return 1
	}
	result := 1.0
	for i := 2; i <= n; i++ {
		result *= float64(i)
	}
	return result
}

// PoissonProbability calculates the Poisson probability P(X=k) for a given lambda.
// P(X=k) = (λ^k * e^-λ) / k!
func PoissonProbability(lambda float64, k int) float64 {
	if k < 0 {
		return 0
	}
	if lambda == 0 {
		if k == 0 {
			return 1 // P(X=0|λ=0) = 1
		}
		return 0
	}
	if lambda < 0 {
		return 0
	}
	return math.Pow(lambda, float64(k)) * math.Exp(-lambda) / Factorial(k)
}

// PoissonPrediction contains the result of a Poisson-based match prediction.
type PoissonPrediction struct {
	ExpectedHomeGoals float64
	ExpectedAwayGoals float64
	HomeWinProb       float64
	DrawProb          float64
	AwayWinProb       float64
	Over25Prob        float64
	Under25Prob       float64
	BTTSProb          float64
	MostLikelyScores  []ScoreProbability
}

// ScoreProbability represents a score and its probability.
type ScoreProbability struct {
	HomeGoals   int
	AwayGoals   int
	Probability float64
}

// CalculatePoissonPrediction computes match outcome probabilities using the Poisson distribution.
// It takes team attack/defense strengths and league average goals to predict outcomes.
func CalculatePoissonPrediction(
	homeGoalsAvg, homeConcededAvg,
	awayGoalsAvg, awayConcededAvg,
	leagueAvgGoals float64,
) PoissonPrediction {
	if leagueAvgGoals == 0 {
		leagueAvgGoals = 2.75 // Default league average
	}

	halfLeagueAvg := leagueAvgGoals / 2

	// Calculate attack and defense strengths
	homeAttackStrength := homeGoalsAvg / halfLeagueAvg
	homeDefenseStrength := homeConcededAvg / halfLeagueAvg
	awayAttackStrength := awayGoalsAvg / halfLeagueAvg
	awayDefenseStrength := awayConcededAvg / halfLeagueAvg

	// Calculate expected goals
	expectedHomeGoals := homeAttackStrength * awayDefenseStrength * halfLeagueAvg
	expectedAwayGoals := awayAttackStrength * homeDefenseStrength * halfLeagueAvg

	// Calculate probabilities using score matrix
	maxGoals := 10
	var homeWinProb, drawProb, awayWinProb float64
	var over25Prob, bttsProb float64

	scores := make([]ScoreProbability, 0)

	for i := 0; i <= maxGoals; i++ {
		for j := 0; j <= maxGoals; j++ {
			prob := PoissonProbability(expectedHomeGoals, i) * PoissonProbability(expectedAwayGoals, j)

			if i > j {
				homeWinProb += prob
			} else if i == j {
				drawProb += prob
			} else {
				awayWinProb += prob
			}

			if i+j > 2 {
				over25Prob += prob
			}
			if i > 0 && j > 0 {
				bttsProb += prob
			}

			if i <= 5 && j <= 5 {
				scores = append(scores, ScoreProbability{
					HomeGoals:   i,
					AwayGoals:   j,
					Probability: prob * 100,
				})
			}
		}
	}

	// Sort scores by probability (descending)
	sortScoresByProbability(scores)

	// Take top 10
	if len(scores) > 10 {
		scores = scores[:10]
	}

	return PoissonPrediction{
		ExpectedHomeGoals: expectedHomeGoals,
		ExpectedAwayGoals: expectedAwayGoals,
		HomeWinProb:       homeWinProb * 100,
		DrawProb:          drawProb * 100,
		AwayWinProb:       awayWinProb * 100,
		Over25Prob:        over25Prob * 100,
		Under25Prob:       (1 - over25Prob) * 100,
		BTTSProb:          bttsProb * 100,
		MostLikelyScores:  scores,
	}
}

// sortScoresByProbability sorts scores by probability in descending order.
func sortScoresByProbability(scores []ScoreProbability) {
	// Simple bubble sort for small arrays
	n := len(scores)
	for i := 0; i < n-1; i++ {
		for j := 0; j < n-i-1; j++ {
			if scores[j].Probability < scores[j+1].Probability {
				scores[j], scores[j+1] = scores[j+1], scores[j]
			}
		}
	}
}

// CalculateGoalProbabilities computes outcome probabilities from expected goals.
func CalculateGoalProbabilities(homeExpectedGoals, awayExpectedGoals float64) (homeWin, draw, awayWin, over25, under25, btts float64) {
	maxGoals := 10

	for i := 0; i <= maxGoals; i++ {
		for j := 0; j <= maxGoals; j++ {
			prob := PoissonProbability(homeExpectedGoals, i) * PoissonProbability(awayExpectedGoals, j)
			if i > j {
				homeWin += prob
			} else if i == j {
				draw += prob
			} else {
				awayWin += prob
			}
			if i+j > 2 {
				over25 += prob
			}
			if i > 0 && j > 0 {
				btts += prob
			}
		}
	}
	under25 = 1 - over25
	return
}

// CalculatePoissonProbabilities returns probability distribution for 0 to maxGoals.
func CalculatePoissonProbabilities(lambda float64, maxGoals int) []float64 {
	probs := make([]float64, maxGoals+1)
	for k := 0; k <= maxGoals; k++ {
		probs[k] = PoissonProbability(lambda, k)
	}
	return probs
}
